### ISSUE TITLE
Exit code on sandbox worker

### ISSUE THREAD

**Issue by PaulGrimshaw** [taskforcesh/bullmq#296]
Is there a way to force a sandbox worker to be killed on error?

I need to ensure that the worker process is properly terminated and not hanging after exit.  

This is the relevant code:

https://github.com/taskforcesh/bullmq/blob/master/src/classes/sandbox.ts

Currently we implement our own try/catch method inside the job runner, and have to terminate with `process.exit()`  which causes this line to run.

```
    exitHandler = (exitCode: any, signal: any) => {
        reject(
          new Error('Unexpected exit code: ' + exitCode + ' signal: ' + signal),
        );
      };
```

I also worry this bypasses all the BullMQ cleanup, but not sure how else to ensure the worker terminates immediately on error?

FYI we have implemented our own locking/timeout behaviour inside the job run (below).  The line I would like to remove is `setTimeout(() => process.exit(), TIMEOUT_RETRY_WAIT_SEC)`. The reason we need it there is the job continues to run after timeout as the sandboxed process is not killed off.

```
  const EventEmitter = require("events")
    EventEmitter.defaultMaxListeners = 50
    process.setMaxListeners(50)

    const { job, redlock, fullQueueName, timeoutSec, jobLogger, connections } = await prepareJobRun(bullJob)
    const lock = await tryGetLock(redlock, `bull:${fullQueueName}_lock`, timeoutSec, jobLogger)

    let result = "OK"
    try {
        await Promise.race([job.run(bullJob.data || {}), jobTimeoutError(timeoutSec * 1000)])
        await lock.unlock()
    } catch (err) {
        setTimeout(() => process.exit(), TIMEOUT_RETRY_WAIT_SEC * 1000)
        if (R.startsWith("Timeout: ", err.message)) await lock.extend(TIMEOUT_RETRY_WAIT_SEC * 1000)

        result = `ERROR: ${err.message}`
        console.error(err)
        jobLogger.error(err)
    } finally {
        await disconnectConnections(connections)
        jobLogger.info({ message: "Job finished", result })
    }
```

**Comment by manast** [comment]
But why can't you just throw an exception from your process handler so that BullMQ handles it as a standard failed job?

**Comment by PaulGrimshaw** [comment]
Thats what we tried originally, but for example running this in the job processor:

```
      setTimeout(() => console.log("LATE log"), 10000)
        throw new Error("Test error")
```

We always see the "LATE LOG" unless we call process.exit() ourselves

**Comment by manast** [comment]
yes, that makes sense since the timeout has not been cancelled. You will need to clean up all running timers. I think it is a better practice to do a cleanup and then a throw than a brute force process exit.

**Comment by PaulGrimshaw** [comment]
I take your point, but this example was trivial - more often it might be because of timeouts connecting to various apis or database connections.  The advantage of the brute force option is that if any error occurs, we know that all existing connections, running queries, etc are all terminated.

**Comment by manast** [comment]
Adding a bit more of information, processes in sandboxed processors are reused, so if you leave dangling timers they will continue working. It is possible to force a process.exit though, that will fail the job but also will imply the creation of a new pristine process for the next jobs to process: https://docs.bullmq.io/patterns/timeout-for-sandboxed-processors
